### DISCUSSION TITLE
[Mac] Homebrew 构建git版mpv

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#70] [mpv 通用]
mac版的mpv在交互上要比win版光秃秃的界面要强一丢丢
![Snipaste_2021-12-23_15-09-06](https://user-images.githubusercontent.com/41094733/147258953-9da481ee-0374-43c2-9394-838440799465.jpg)

# 使用deus0ww维护的分支仓库安装mpv

m2pro Sonoma 的安装记录。

### 1. 前置工作

安装商店的xcode15，并执行一次初次启动后关闭。

安装Homebrew
https://brew.sh
_更新homebrew自身_
`brew update`

打开终端，安装gcc
`brew install gcc`

安装py的必要依赖
`brew install python-setuptools`

### 2. 安装流程

以下命令安装deus0ww构建版 https://github.com/deus0ww/homebrew-tap
`brew tap deus0ww/homebrew-tap`

对于使用最新的vk后端，这是必要的
`brew install molten-vk`

（可选）网络流播放功能
`brew install yt-dlp`

（可选）vapoursynth滤镜功能
`brew install vapoursynth`

> 额外建议先安装git版的部分依赖来满足最新mpv的特性需要
> `brew install --HEAD deus0ww/homebrew-tap/libass`
> `brew install --HEAD deus0ww/homebrew-tap/libplacebo`
> `brew install --HEAD deus0ww/homebrew-tap/ffmpeg`

`brew install --HEAD deus0ww/homebrew-tap/mpv`

### 3. 后续处理

安装完成后bundle应该位于以下目录
intel `/usr/local/Cellar/mpv/HEAD-*******/mpv.app`
apple silicon `/opt/homebrew/Cellar/mpv/HEAD-*******/mpv.app`
自行软链至`应用程序`目录

主设置文件的路径位于
`.config/mpv/`

### 4. 更新

首先更新brew自身

`brew update`

倒序卸载。当前mpv的新特性和libplacebo强相关，因此每次更新都建议一起更新依赖

`brew uninstall mpv ffmpeg libplacebo libass`

然后再次运行前文第二部分的几个命令

**Comment by hooke007** [comment]
尝试最新的vulkan后端

## mpv.conf

```ini
 vo = gpu-next                # 使用新渲染器
 gpu-context = macvk
 hwdec = videotoolbox
 target-colorspace-hint = auto # 实测支持自动激活hdr输出（需要gpu-next）

 autofit-smaller = 50%x50%
 autofit-larger = 100%x100%
 hidpi-window-scale = no
```

**Comment by eko5624** [comment]
我弄了个github action版 
https://github.com/eko5624/mpv-macos-test

**Reply by hooke007** [reply]
我也没有使用homebrew cask版的原因是不支持vapoursynth、
你有无测试过只用bundle能不能跑vs（应该是不能的缺少依赖似乎没发打包进包里

**Reply by hooke007** [reply]
而且编译mpv还有另外的几个用途，比如svp的iina支持

**Reply by eko5624** [reply]
vapoursynth没在mac上用过。我在mac上仅仅用了mpv的播放视频功能。vapoursynth的用法还没研究过。

**Comment by eko5624** [comment]
我刚看了一下deus0ww编译生成的bundle包里面有2个vapoursynth相关文件，不知道管不管用。
![F0851B2334973E90C662BB1F151D7A5D](https://user-images.githubusercontent.com/84926938/147375565-a7504b58-6246-4ed1-8796-aafb218a90bf.jpg)

**Reply by hooke007** [reply]
这两应该只是提供了vf=vapoursynth的滤镜功能接口

应该还是逃不开外部自行编译vs和特定版本的python（

**Reply by hooke007** [reply]
其实很方便测试的，如果你之前没用homebrew装过vs和py的话
打开一个视频然后跑一个空脚本看看控制台有无报错就行

test.vpy
```
import vapoursynth as vs
core = vs.core
clip = video_in
clip.set_output()
```

**Reply by eko5624** [reply]
具体怎么跑的呢？Mpv打开视频，然后这个脚本怎么用呢。

**Reply by eko5624** [reply]
我测试了一下，似乎没有报错，你看一下。
MPV用的是github action版
![4246C08EE563C1DF7F3AD12A1A77969C](https://user-images.githubusercontent.com/84926938/147376782-f087a9f8-160c-424a-82af-6849120059dd.jpg)

**Reply by eko5624** [reply]
你也可以下载我这个github action版测试一下

**Reply by hooke007** [reply]
你的系统应该有vs，我这里移除homebrew的vs再跑脚本就直接无法初始化报错了

移除前
<img width="590" alt="截屏2021-12-25 上午4 33 17" src="https://user-images.githubusercontent.com/41094733/147377456-6ef7114f-d320-4686-8365-18e8b9f76c84.png">

移除后
<img width="590" alt="截屏2021-12-25 上午4 36 31" src="https://user-images.githubusercontent.com/41094733/147377490-0f7cefc3-aeac-418f-b5f6-35b783dcaeb2.png">

**Reply by eko5624** [reply]
确实。我系统里之前用brew装过vapoursynth，uninstall后就报错了。

**Comment by nyfair** [comment]
还是直接用iina的编译脚本加上vapoursynth依赖吧，不过水果祖传opengl4.1都9年没更新了真的建议用mpv麽？

**Reply by hooke007** [reply]
mpv在我mac上只是备用测试的，效能实在太差劲了

**Comment by hooke007** [comment]
mac下真的离了大谱，软硬解tonemapping出来的颜色差异巨大（看着像动态的问题），更要被我扫进软件垃圾堆了

**Comment by nyfair** [comment]
mac没有vulkan，opengl都是远古版本，跑不了libplacebo，mpv团队又没意愿更新老的vo，不就是没戏麽。至于某些野生补丁用molten_vk强行模拟vulkan，和那个用opengles模拟directx的angle一样都属于没啥实用性的自娱自乐产物

hooke007 ***@***.***> 于 2022年1月3日周一 下午11:05写道：

**Comment by eko5624** [comment]
我又搞了一个在mac平台静态编译的FFmpeg和MPV。应该intel平台和apple silicon都支持。我没有silicon平台的机器，暂时无法测试和完善silicon平台。目前FFmpeg可以完全实现静态编译, MPV除了luajit和molten-vk是动态链接，其它的包都是静态链接。仓库地址：
https://github.com/eko5624/mpv-mac

**Reply by EraSun** [reply]
被iina的HDR BUG搞得烦死了，谢谢你，这个我在m2 macmini上面使用暂时没有出现什么问题

**Reply by nyfair** [reply]
luajit静态链接的开关在https://repo.or.cz/luajit-2.0.git/blob/HEAD:/src/Makefile#l73
章亦春跑去修炼入定之后，早就不维护openresty那个luajit fork了，加上mpv的需求也和nginx不一样，看不出使用那个fork的意义

另外建议加上vapoursynth，homebrew可以抄的https://github.com/Homebrew/homebrew-core/blob/master/Formula/v/vapoursynth.rb

**Reply by eko5624** [reply]
vapoursynth是不是不能静态链接至ffmpeg？

**Reply by nyfair** [reply]
不能吧，运行模式决定的，除非把python也静态链接进去

**Reply by eko5624** [reply]
尝试静态编译luajit，mpv也可以编译成功。但双击打开mpv.app会闪退。命令行打开mpv会提示出错信息：
```
PANIC: unprotected error in call to Lua API (8)
```

**Reply by nyfair** [reply]
https://github.com/eko5624/mpv-mac/releases/download/dev/luajit2.tar.xz
pkgconfig里
Libs.private: -lm -ldl
改成
Libs.private: -Wl,-E -lm -ldl

**Reply by eko5624** [reply]
> https://github.com/eko5624/mpv-mac/releases/download/dev/luajit2.tar.xz pkgconfig里 Libs.private: -lm -ldl 改成 Libs.private: -Wl,-E -lm -ldl

编译mpv时会出错，提示：
```
ld: unknown option: -E
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Reply by eko5624** [reply]
> 另外建议加上vapoursynth，homebrew可以抄的https://github.com/Homebrew/homebrew-core/blob/master/Formula/v/vapoursynth.rb

已加！
https://github.com/eko5624/mpv-mac/releases/tag/2024-07-20

**Comment by eko5624** [comment]
@hooke007 @EraSun 现已支持原生arm64平台的mac。我没有M平台的设备，可以帮忙测试一下下面的arm64版本是否运行良好？
https://github.com/eko5624/mpv-mac/releases

**Reply by hooke007** [reply]
已损坏无法运行

**Reply by eko5624** [reply]
命令行运行有啥报错信息吗？再帮忙测下arm64版的ffmpeg能运行吗？

**Reply by hooke007** [reply]
你压缩包里那个不是bundle吗，哪来的命令行版本

ffmpeg粗看没啥问题

**Reply by hooke007** [reply]
拆包运行的结果
```
hooke007@hooke007-mbp-16-m2pro ~ % /Users/hooke007/Downloads/mpv/mpv.app/Contents/MacOS/mpv ; exit;
dyld[61960]: symbol not found in flat namespace '_OBJC_CLASS_$__TtC5swift11Application'
zsh: abort      /Users/hooke007/Downloads/mpv/mpv.app/Contents/MacOS/mpv

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.
```

**Reply by eko5624** [reply]
好的，明天我研究下是哪里的问题。

**Reply by eko5624** [reply]
手误，这里的swiftflags设置错误，已修正。
https://github.com/eko5624/mpv-mac/blob/8b491510a03a68b574c5a05d7fc37f8af918c8f3/scripts/mpv.sh#L23

下面这版应该没问题了：
https://github.com/eko5624/mpv-mac/releases/download/2024-07-24/mpv-arm64-git-cb75ecf19f.zip
